### PR TITLE
feat: implement gateway middleware

### PR DESCRIPTION
**Pull request by obdulia-losantos**
- [x] Checks `X-Oauth-Token` header to find out current user
- [x] Enables two url formats `/{realm}/{app}/{endpoint}` (behind gateway) and `/{endpoint}` (internal calls)
- [x] Enables custom urls for:
  - admin `/-/{app}/admin/`
  - auth `/-/{app}/accounts/`
  - static `/-/{app}/static/` 
- [x] Serializer url fields shows the Gateway url if present.
- [x] `/-/{app}/accounts/logout/` redirects to kong logout endpoint
- [x] Added `/logout` endpoint to be used in UI.
- [x] UI uses full path to build AJAX api calls.
- [x] UI `get_kernel_url` view should return the external URL with the realm path if necessary. (**Enhacement**)
- [x] Test, test, test

### How to check this PR

In `aether`

1. Build `ui-assets` `docker-compose build ui-assets && docker-compose run ui-assets build`

In `aether-bootstrap`

2. Clone `aether-bootstrap` and use this branch https://github.com/eHealthAfrica/aether-bootstrap/tree/feat/kong-keycloak

3. Change `docker-compose-base.yml` to point to these images (something like the following)

```yml
kernel-base:
    # image: ehealthafrica/aether-kernel:${AETHER_VERSION}
    image: aether-kernel
    build: ${PATH_TO_LOCAL_AETHER}/aether-kernel
    volumes:
      - ${PATH_TO_LOCAL_AETHER}/aether-kernel:/code
      # common folder
      - ${PATH_TO_LOCAL_AETHER}/aether-common-library/aether/common:/code/aether/common
      # ...
```
Same for `odk` and `ui`

4. Change the init script (`scripts/initialise_docker_environment.sh`) to not to pull aether images
https://github.com/eHealthAfrica/aether-bootstrap/blob/feat/kong-keycloak/scripts/initialise_docker_environment.sh#L52

```bash
echo "${LINE} Preparing aether containers..."
# setup container (model migration, admin user, static content...)
CONTAINERS=( kernel ui odk )
for container in "${CONTAINERS[@]}"
do
    # docker-compose pull $container
    docker-compose run --no-deps $container setup
done
```
5. Remove `.env` file
6. Execute init script `./scripts/initialise_docker_environment.sh`
7. Start auxiliary containers `docker-compose up -d db minio kong keycloak`
8. Start aether containers `docker-compose up kernel odk ui`
9. Go to `http://aether.local/dev/ui/` or `http://aether.local/prod/ui/`

### CODE REVIEW

File: aether-common-library/aether/common/tests/test_urls.py [hunk 1]
```diff
@@ -80,6 +89,7 @@ def test__urls__accounts(self):
 @override_settings(
     CAS_SERVER_URL='http://localhost:6666',
```
Comment (shawnsarwar) [inline aether-common-library/aether/common/tests/test_urls.py:93]: Are you calling the CAS_SERVER evil?

**Review by shawnsarwar — DISMISSED** [review]
Looks great! Let's get this sucker RC'd

File: aether-common-library/aether/common/tests/test_urls.py (hunk 1 above)
Comment (obdulia-losantos) [reply aether-common-library/aether/common/tests/test_urls.py:93]: :smiling_imp:

File: aether-common-library/aether/common/conf/urls.py [hunk 2]
```diff
@@ -90,67 +161,53 @@ def generate_urlpatterns(token=False, kernel=False, app=[]):
...
-    if settings.DEBUG and 'debug_toolbar' in settings.INSTALLED_APPS:  # pragma: no cover
-        import debug_toolbar
+def __check_kernel_env():
+    from aether.common.kernel.utils import get_kernel_server_url, get_kernel_server_token
 
-        urlpatterns += [
-            path(route='__debug__/', view=include(debug_toolbar.urls)),
-        ]
+    # `aether.common.kernel.utils.get_kernel_server_url()` returns different
+    #  values depending on the value of `settings.TESTING`. Without these
```
Comment (admbtlr) [inline aether-common-library/aether/common/conf/urls.py]:
```suggestion
    # values depending on the value of `settings.TESTING`. Without these
```

File: aether-common-library/aether/common/utils.py [hunk 3]
```diff
@@ -58,23 +58,32 @@ def find_in_request(request, key, default_value=None):
...
+    Finds the key in within the request headers.
+
     https://docs.djangoproject.com/en/2.2/ref/request-response/#django.http.HttpRequest.META
 
     New in Django.2.2
     https://docs.djangoproject.com/en/2.2/ref/request-response/#django.http.HttpRequest.headers
     '''
 
-    return getattr(request, 'session', {}).get(
+    return getattr(request, 'headers', {}).get(  # New in Django.2.2
```
Comment (admbtlr) [inline aether-common-library/aether/common/utils.py]:
```suggestion
    return getattr(request, 'headers', {}).get(  # New in Django 2.2
```

**Review by admbtlr — CHANGES_REQUESTED** [review]
😍 😍 😍 😍 😍 😍 

Just a couple of VERY IMPORTANT CHANGES

**Review by admbtlr — APPROVED** [review]

**Pull request merged**